### PR TITLE
Don't modify the on disk cache in fine-grained mode

### PR DESCRIPTION
**Pull request by msullivan**
This is a little subtle, because interface_hash still needs to be
computed, as it is a major driver of the coarse-grained build process.

Since metas are no longer computed for files that get rechecked during
build, to avoid spuriously reprocessing them we need to find initial
file state in cache mode as well.

### CODE REVIEW

File: mypy/dmypy_server.py [hunk 1]
```diff
@@ -254,9 +254,8 @@ def initialize_fine_grained(self, sources: List[mypy.build.BuildSource]) -> Dict
         self.fscache = FileSystemCache(self.options.python_version)
         self.fswatcher = FileSystemWatcher(self.fscache)
         self.update_sources(sources)
-        if not self.options.use_fine_grained_cache:
-            # Stores the initial state of sources as a side effect.
-            self.fswatcher.find_changed()
+        # Stores the initial state of sources as a side effect.
```
Comment (JukkaL) [inline mypy/dmypy_server.py]: Why was this changed to be executed unconditionally?

File: mypy/dmypy_server.py [hunk 2]
```diff
@@ -274,16 +273,17 @@ def initialize_fine_grained(self, sources: List[mypy.build.BuildSource]) -> Dict
...
+        # If we are using the fine-grained cache, build might not have
+        # actually done the typechecking on the updated files yet.
         # Run a fine-grained update starting from the cached data
         if self.options.use_fine_grained_cache:
             # Pull times and hashes out of the saved_cache and stick them into
             # the fswatcher, so we pick up the changes.
             for state in self.fine_grained_manager.graph.values():
                 meta = state.meta
+                # If there isn't a meta, that means the current
+                # version got checked in the initial build.
```
Comment (JukkaL) [inline mypy/dmypy_server.py]: Relying on this logic here seems a bit fragile. Maybe move the `meta is None` check to a helper method in `State` and use the helper method here instead?

File: mypy/dmypy_server.py (hunk 1 above)
Comment (msullivan) [reply mypy/dmypy_server.py]: So that we have accurate `fswatcher` cache information for files that we didn't read from the on-disk cache, now that we don't generate `CacheMeta`s

**Comment by msullivan** [comment]
The main rationale is closer to correctness than performance. Currently we write and delete cache files during the initial load of fine-grained mode, but never during fine-grained updates. This means that fine-grained can mess the cache up but won't fix it without a restart, which seems not great.

File: mypy/dmypy_server.py (hunk 2 above)
Comment (msullivan) [reply mypy/dmypy_server.py]: Changing it to use `is_cache_skeleton` instead.

**Comment by msullivan** [comment]
This has a bad interaction with #4669 and needs changes. My plan is to land #4669 and revise this after.
